### ISSUE TITLE
ConcurrentModificationException while saving

### ISSUE THREAD

**Issue by flori-schwa** [flori-schwa/VarLightOld#10]
[Permalink to spigot post](https://www.spigotmc.org/threads/varlight.362135/page-6#post-3598225)

[VarLight Version 1.4.5](https://github.com/flori-schwa/VarLight/tree/1.4.5)

Using Built-in methods (no LightAPI)

Stacktrace
```
[21:26:44] [Craft Scheduler Thread - 288/WARN]: [VarLight] Plugin VarLight v1.4.5 generated an exception while executing task 7
java.util.ConcurrentModificationException: null
        at java.util.HashMap$HashIterator.nextNode(Unknown Source) ~[?:1.8.0_231]
        at java.util.HashMap$ValueIterator.next(Unknown Source) ~[?:1.8.0_231]
        at me.shawlaf.varlight.persistence.LightSourcePersistor.save(Unknown Source) ~[?:?]
        at me.shawlaf.varlight.VarLightPlugin.lambda$null$0(Unknown Source) ~[?:?]
        at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(Unknown Source) ~[?:1.8.0_231]
        at java.util.stream.ReferencePipeline$3$1.accept(Unknown Source) ~[?:1.8.0_231]
        at java.util.stream.ReferencePipeline$2$1.accept(Unknown Source) ~[?:1.8.0_231]
        at java.util.stream.ReferencePipeline$3$1.accept(Unknown Source) ~[?:1.8.0_231]
        at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(Unknown Source) ~[?:1.8.0_231]
        at java.util.stream.AbstractPipeline.copyInto(Unknown Source) ~[?:1.8.0_231]
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source) ~[?:1.8.0_231]
        at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(Unknown Source) ~[?:1.8.0_231]
        at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(Unknown Source) ~[?:1.8.0_231]
        at java.util.stream.AbstractPipeline.evaluate(Unknown Source) ~[?:1.8.0_231]
        at java.util.stream.ReferencePipeline.forEach(Unknown Source) ~[?:1.8.0_231]
        at me.shawlaf.varlight.VarLightPlugin.lambda$initAutosave$1(Unknown Source) ~[?:?]
        at org.bukkit.craftbukkit.v1_14_R1.scheduler.CraftTask.run(CraftTask.java:81) ~[spigot.jar:git-Spigot-94af569-1e7a197]
        at org.bukkit.craftbukkit.v1_14_R1.scheduler.CraftAsyncTask.run(CraftAsyncTask.java:54) [spigot.jar:git-Spigot-94af569-1e7a197]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:1.8.0_231]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:1.8.0_231]
        at java.lang.Thread.run(Unknown Source) [?:1.8.0_231]
Suppressed: java.io.IOException: Incomplete document
        at com.google.gson.stream.JsonWriter.close(JsonWriter.java:559) ~[spigot.jar:git-Spigot-94af569-1e7a197]
        at me.shawlaf.varlight.persistence.LightSourcePersistor.save(Unknown Source) ~[?:?]
        at me.shawlaf.varlight.VarLightPlugin.lambda$null$0(Unknown Source) ~[?:?]
        at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(Unknown Source) ~[?:1.8.0_231]
        at java.util.stream.ReferencePipeline$3$1.accept(Unknown Source) ~[?:1.8.0_231]
        at java.util.stream.ReferencePipeline$2$1.accept(Unknown Source) ~[?:1.8.0_231]
        at java.util.stream.ReferencePipeline$3$1.accept(Unknown Source) ~[?:1.8.0_231]
        at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(Unknown Source) ~[?:1.8.0_231]
        at java.util.stream.AbstractPipeline.copyInto(Unknown Source) ~[?:1.8.0_231]
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source) ~[?:1.8.0_231]
        at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(Unknown Source) ~[?:1.8.0_231]
        at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(Unknown Source) ~[?:1.8.0_231]
        at java.util.stream.AbstractPipeline.evaluate(Unknown Source) ~[?:1.8.0_231]
        at java.util.stream.ReferencePipeline.forEach(Unknown Source) ~[?:1.8.0_231]
        at me.shawlaf.varlight.VarLightPlugin.lambda$initAutosave$1(Unknown Source) ~[?:?]
        at org.bukkit.craftbukkit.v1_14_R1.scheduler.CraftTask.run(CraftTask.java:81) ~[spigot.jar:git-Spigot-94af569-1e7a197]
        at org.bukkit.craftbukkit.v1_14_R1.scheduler.CraftAsyncTask.run(CraftAsyncTask.java:54) [spigot.jar:git-Spigot-94af569-1e7a197]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:1.8.0_231]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:1.8.0_231]
        at java.lang.Thread.run(Unknown Source) [?:1.8.0_231]
```

**Comment by SlimeDog** [comment]
It's been 7 days. Any issues with this solution?

**Comment by flori-schwa** [comment]
There's been some progress on the vldb branch: https://github.com/flori-schwa/VarLight/commits/vldb

**Comment by SlimeDog** [comment]
Thanks. Release prognosis?

**Comment by flori-schwa** [comment]
I'll get a pre-release ready as soon as The Rest of [VLDBFile](https://github.com/flori-schwa/VarLight/blob/vldb/Core/src/me/shawlaf/varlight/persistence/vldb/VLDBFile.java) is implemented, [LightSourcePersistor](https://github.com/flori-schwa/VarLight/blob/vldb/Core/src/me/shawlaf/varlight/persistence/LightSourcePersistor.java) uses that implementation and perhaps the Light update code is improved

**Comment by SlimeDog** [comment]
Excellent

**Comment by flori-schwa** [comment]
This should hopefully no longer occur in the [Latest pre-release](https://github.com/flori-schwa/VarLight/releases/tag/pre3-1.4.6-VLDB)

**Comment by flori-schwa** [comment]
[Should hopefully be fixed in 1.4.6](https://github.com/flori-schwa/VarLight/releases/tag/1.4.6)

**Comment by SlimeDog** [comment]
Looks to be fixed. Thanks.
